### PR TITLE
APEXMALHAR-1957: Added threading for reading data from hbase. Added s…

### PR DESCRIPTION
**Pull request by bhupeshchawda**
…upport for progressive read. Added unit tests.

### CODE REVIEW

File: contrib/src/main/java/com/datatorrent/contrib/hbase/HBasePOJOInputOperator.java [hunk 1]
```diff
@@ -69,116 +69,103 @@ public Object convertValue( HBaseFieldInfo fieldInfo, Object value)
...
+    } else {
+      // If end row also specified
+      scan = new Scan(Bytes.toBytes(getLastReadRow() == null ? getStartRow() : getLastReadRow()),
+          Bytes.toBytes(getEndRow()));
+    }
+    for (HBaseFieldInfo field : tableInfo.getFieldsInfo()) {
+      scan.addColumn(Bytes.toBytes(field.getFamilyName()), Bytes.toBytes(field.getColumnName()));
+    }
+    scan.setAttribute(Scan.HINT_LOOKAHEAD, Bytes.toBytes(getHintScanLookahead()));
+    return scan;
```
Comment (sandeepdeshmukh) [inline contrib/src/main/java/com/datatorrent/contrib/hbase/HBasePOJOInputOperator.java:133]: Can you reuse the scan variable? That will save connection time for every request.

File: contrib/src/main/java/com/datatorrent/contrib/hbase/HBasePOJOInputOperator.java [hunk 2]
```diff
@@ -69,116 +69,103 @@ public Object convertValue( HBaseFieldInfo fieldInfo, Object value)
...
-  }
-  public void setStore(HBaseStore store)
+  @Override
+  protected Scan operationScan()
   {
-    this.store = store;
+    Scan scan;
+    if (getLastReadRow() == null && getStartRow() == null) {
+      // If no start row specified and no row read yet
+      scan = new Scan();
```
Comment (sandeepdeshmukh) [inline contrib/src/main/java/com/datatorrent/contrib/hbase/HBasePOJOInputOperator.java]: new scan() can be moved to setup()

File: contrib/src/main/java/com/datatorrent/contrib/hbase/HBaseScanOperator.java [hunk 3]
```diff
@@ -40,27 +51,106 @@
...
+  {
+    super.beginWindow(windowId);
+    tuplesRead = 0;
+  }
 
   @Override
   public void emitTuples()
   {
+    if (!readThread.isAlive() && resultQueue.isEmpty()) {
+      startReadThread();
```
Comment (sandeepdeshmukh) [inline contrib/src/main/java/com/datatorrent/contrib/hbase/HBaseScanOperator.java]: Better to throw RuntimeException and redeploy the operator.

File: contrib/src/main/java/com/datatorrent/contrib/hbase/HBaseScanOperator.java [hunk 4]
```diff
@@ -40,27 +51,106 @@
...
-      Scan scan = operationScan();
-      ResultScanner scanner = table.getScanner(scan);
-      for (Result result : scanner) {
-        //KeyValue[] kvs = result.raw();
-        //T t = getTuple(kvs);
-        T t = getTuple(result);
-        outputPort.emit(t);
+      Result result = resultQueue.poll();
+      if (result == null) {
+        Thread.sleep(sleepMillis);
```
Comment (sandeepdeshmukh) [inline contrib/src/main/java/com/datatorrent/contrib/hbase/HBaseScanOperator.java]: Need not sleep here as the engine will handle it if there is no tuple emitted during the call of emitTuples.

File: contrib/src/main/java/com/datatorrent/contrib/hbase/HBaseScanOperator.java [hunk 5]
```diff
@@ -40,27 +51,106 @@
...
+        return;
+      }
+      T tuple = getTuple(result);
+      if (tuple != null) {
+        outputPort.emit(tuple);
+        tuplesRead++;
       }
     } catch (Exception e) {
       e.printStackTrace();
```
Comment (sandeepdeshmukh) [inline contrib/src/main/java/com/datatorrent/contrib/hbase/HBaseScanOperator.java]: Throw RTE.

File: contrib/src/test/java/com/datatorrent/contrib/hbase/HBasePOJOInputOperatorTest.java [hunk 6]
```diff
@@ -128,7 +144,7 @@ public void populateDAG(DAG dag, Configuration conf)
         Thread.sleep(1000);
       }
       catch( Exception e ){}
-      
+      logger.info("Tuple row key: ", output.getReceivedTuples());
       logger.info( "Received tuple number {}, instance is {}.", output.getReceivedTuples() == null ? 0 : output.getReceivedTuples().size(), System.identityHashCode( output ) );
       if( output.getReceivedTuples() != null && output.getReceivedTuples().size() == TUPLE_NUM )
```
Comment (sandeepdeshmukh) [inline contrib/src/test/java/com/datatorrent/contrib/hbase/HBasePOJOInputOperatorTest.java]: If this condition is never met, the unit test case will run forever. Need some timeout based failure of the test case.

File: contrib/src/main/java/com/datatorrent/contrib/hbase/HBasePOJOInputOperator.java (hunk 1 above)
Comment (bhupeshchawda) [inline contrib/src/main/java/com/datatorrent/contrib/hbase/HBasePOJOInputOperator.java:133]: done

File: contrib/src/main/java/com/datatorrent/contrib/hbase/HBaseScanOperator.java (hunk 3 above)
Comment (bhupeshchawda) [inline contrib/src/main/java/com/datatorrent/contrib/hbase/HBaseScanOperator.java]: done

File: contrib/src/main/java/com/datatorrent/contrib/hbase/HBaseScanOperator.java (hunk 5 above)
Comment (bhupeshchawda) [inline contrib/src/main/java/com/datatorrent/contrib/hbase/HBaseScanOperator.java]: done

File: contrib/src/main/java/com/datatorrent/contrib/hbase/HBaseScanOperator.java (hunk 4 above)
Comment (bhupeshchawda) [inline contrib/src/main/java/com/datatorrent/contrib/hbase/HBaseScanOperator.java]: done

File: contrib/src/test/java/com/datatorrent/contrib/hbase/HBasePOJOInputOperatorTest.java (hunk 6 above)
Comment (bhupeshchawda) [inline contrib/src/test/java/com/datatorrent/contrib/hbase/HBasePOJOInputOperatorTest.java]: done. Added a check for 30 seconds.

**Comment by bhupeshchawda** [comment]
Addressed review comments.

**Comment by bhupeshchawda** [comment]
Closing this PR due to improper branch name. Will reopen another one with correct one.

**Pull request closed without merging**